flink-connector/src/main/java/org/apache/flink/connector/delta/sink/committer/DeltaGlobalCommitter.java

```diff
@@ -49,27 +73,362 @@
+     *       is performed. Such situations may happen when e.g. there were no stream events received
+     *       within given checkpoint interval,
+     *   <li>If appId is successfully resolved then the provided set of committables needs to be
+     *       flattened (as one {@link DeltaGlobalCommittable} contains a list of
+     *       {@link DeltaCommittable}), mapped to {@link AddFile} objects and then grouped by
+     *       checkpointId. The grouping part is necessary as committer object may receive
+     *       committables from different checkpoint intervals,
+     *   <li>We process each of the resolved checkpointId in increasing order,
+     *   <li>During processing each of the checkpointId and their committables, we first query
+     *       the DeltaLog for last committed transaction version for given appId. Here transaction
+     *       version equals checkpointId. We proceed with the transaction only if current
+     *       checkpointId is greater than last committed transaction version.
+     *   <li>If above condition is met, then we handle the metadata for data in given stream by
+     *       comparing the stream's schema with current table snapshot's schema. We proceed with
+     *       the transaction only when the schemas are matching or when it was explicitly configured
+     *       during creation of the sink that we can try to update the schema.
+     *   <li>If above validation passes then we prepare the final set of {@link Action} objects to
+     *       be committed along with transaction's metadata and mandatory parameters,
+     *   <li>We try to commit the prepared transaction
+     *   <li>If the commit fails then we fail the application as well. If it succeeds then we
+     *       proceed with the next checkpointId (if any).
+     * </ol>
+     *
+     * @param globalCommittables list of combined committables objects
+     * @return always empty collection as we do not want any retry behaviour
+     */
     @Override
     public List<DeltaGlobalCommittable> commit(List<DeltaGlobalCommittable> globalCommittables) {
+        String appId = resolveAppId(globalCommittables);
+        if (appId != null) { // means there are committables to process
+            SortedMap<Long, List<DeltaCommittable>> committablesPerCheckpoint =
+                groupCommittablesByCheckpointInterval(globalCommittables);
+            DeltaLog deltaLog = DeltaLog.forTable(conf, basePath.getPath());
+
+            for (long checkpointId : committablesPerCheckpoint.keySet()) {
+                OptimisticTransaction transaction = deltaLog.startTransaction();
+                long lastCommittedVersion = transaction.txnVersion(appId);
+                if (checkpointId > lastCommittedVersion) {
+                    doCommit(
+                        transaction,
+                        committablesPerCheckpoint.get(checkpointId),
+                        deltaLog.snapshot().getVersion());
```

Comment:
@scottsand-db thank you for noticing it because what i've missed after last refactor is that i do not really care for current version here and pass only to later check for condition `(currentTableVersion == -1)`. So i changing it from passing `deltaLog.snapshot().getVersion()` to `deltaLog.tableExists()`.

As for the DSW API it seems that it may be a valid extension and simplification for connectors developers.